### ISSUE TITLE
Allow setters to declare a return type of `Never`

### ISSUE THREAD

**Issue by mateusfccp** [dart-lang/language#4202]
This may be only useful on some few edge cases, so I am not sure if this is viable or not, but it may not be so hard to do and isn't breaking.

Currently, the specification explicitly states:

> It is a compile-time error if a setter declares a return type other than void.

With null-safety and the `Never` type introduced, there are cases where one might want to change their setter return type to indicate that an error is never thrown.

For instance, [`UnmodifiableListView.operator []=`](https://api.dart.dev/stable/3.5.3/dart-collection/UnmodifiableListView/operator_put.html) always throw.

**Comment by lrhn** [comment]
The return type of a setter doesn't matter, because the type isn't used anywhere in the specification.
Any returned value is lost, so it's type doesn't matter.

If we allow `Never` as return type, as a signal that the setter invocation never returns any value, then we'd want the spec to recognize that.
That would mean something like: 

- If an assignment expression invokes a setter (it's not an assignment to a local variable), then let _R_ be the return type of the signature of the setter being invoked, and let _S_ be the static type of the assigned expression. If _R_ is a bottom type (any subtype of `Never`), then the static type of the assignment expression is `Never`. Otherwise the static type of the assignment expression is _S_.
- It is be a compile time error if the declared return type of a setter is anything except `void` or `Never`. A setter declaration with no return type defaults to `void` (even if we can see that the code definitely throws). An instance setter with no declared return type inherits a return type from any superinterface member signature as normal, and defaults to `void` if there is no superinterface setter. 
- It's a compile-time error if a concrete setter declaration has a bottom type as return type, and control flow can exit the setter body (reach the end, reach any return statement, or even contain any return statement, but this is just the same rules as for any method body with a bottom return type).
- `operator []=` counts as a setter. 

Could work. 
Not sure if it's worth the work, just to document that a setter throws.
For a static setter, it just means that you should never call it, so it might as well not be there. 
This only makes sense for instance setters, but those also have to be valid overrides, so any interface having a throwing setter implies that all subtypes must also throw.
That again only makes sense for a sub-interface that disables supertype APIs, which isn't great OO design (breaks subtype substitutability). Can be useful occasionally, but not great.

**Comment by mateusfccp** [comment]
Yes, it would only be useful for instance setters, and only for a few cases. I feel like there are no downsides (except, obviously, for the resources spent to implement it), but maybe the benefits are not worthy enough.

I was thinking `operator []=` was already a setter (because it already doesn't allow anything but `void`). Does it have a different special case for the return type?

**Comment by sriramsowmithri9807** [comment]
Hi @mateusfccp 

I have a quick idea to share! Currently, setters can only return void, but with null-safety and the Never type, it might be cool to allow setters to return Never in certain cases.

For example,` UnmodifiableListView.operator []= always throws`, so having a Never return type could clarify intent and enhance type safety.

What do you think? 

Cheers,
Sowmithri Sriram.
sriramsowmithri9807 --> github

**Comment by lrhn** [comment]
> I was thinking `operator []=` was already a setter 

I think it is, just wanted to be certain that it was included since "setter" sometimes means "something declared using `set`" and  sometimes means "something behaving as an assignment". Here we want the latter.

**Comment by rrousselGit** [comment]
One workaround is to mark such setter with a `@Deprecated('do not use')`, like:

```dart
@Deprecated('The list is immutable, do not use');
@override
void operator[]=(int index);
```

**Comment by mateusfccp** [comment]
> One workaround is to mark such setter with a `@Deprecated('do not use')`, like:
> 
> ```dart
> @Deprecated('The list is immutable, do not use');
> @override
> void operator[]=(int index);
> ```

This is an interesting workaround and may work, although the semantics conveyed are not correct.

**Comment by sriramsowmithri9807** [comment]
I agree that allowing setters to use `Never` as a return type would enhance clarity in cases like `UnmodifiableListView.operator []=`. It makes the intent explicit that such setters always throw and aligns well with null-safety principles. This seems like a reasonable and non-breaking improvement."


```dart
class UnmodifiableListView<T> {
  List<T> _items;

  UnmodifiableListView(this._items);

  Never operator []=(int index, T value) {
    throw UnsupportedError('Cannot modify an unmodifiable list.');
  }
}

void main() {
  var list = UnmodifiableListView<int>([1, 2, 3]);

  // Trying to assign a value will always throw an error
  list[0] = 10; // Throws UnsupportedError
}
```

**Comment by Wdestroier** [comment]
The [`@alwaysThrows`](https://api.flutter.dev/flutter/meta/alwaysThrows-constant.html) annotation exists, but it's deprecated:
```dart
@Deprecated("Use a return type of 'Never' instead")
_AlwaysThrows const alwaysThrows
```

**Comment by lrhn** [comment]
I won't expect this to happen. 
It'll require some work, with very, very little _potential_ benefit, and no guaranteed benefit.

If a setter *always throws*, then you should *never assign it*. (Delibarately doing `alwaysThrows = 42` is silly, just `throw` directly, or have a *proper* throwing method.)
If you don't use the setter, the return type doesn't matter.

Making the setter have return type `Never` will only benefit you if you try to assign to something that is (hopefully) documented as "not assignable". And only if you have more code after it that becomes dead code. If the assignment is the last statement of a block, you probably won't get any warning anyway. And then you have to find the throw using your tests.
A `Never` return type doesn't make invoking it an error, it only *may* induce a warning if it introduces dead code,
which it may not do. A "@DoNotCall()" annotation would be more accurate than a `Never` return type.

Without a `Never` return type, your tests should also catch that the invocation throws, likely an `UnsuppportedError` with the name of the member. Then you remove the assignment and the return type again wouldn't do anything.

That is:
- A `Never` return type won't actually prevent invocation, it'll just affect control flow after it.
- If that doesn't dominate other code, which becomes dead, there is _no visible sign_ that it always throws.
- And then you have to catch the throwing in your tests anyway.
- So you migh as well just do that, and not spend effort on a feature with so little benefit.

I'm going to close this as "not planned". Unless something significantly changes the cost/benefit ratio, I don't see this ever becoming the best use of resources.
